compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt

```diff
@@ -413,9 +413,18 @@ class MethodInliner(
 
         val capturedParamsSize = parameters.capturedParametersSizeOnStack
         val realParametersSize = parameters.realParametersSizeOnStack
+        val newArgumentList = if (node != inliningContext.lambdaInfo?.node?.node) {
```

Comment:
I would prefer to make smth like this:
```
        val isInsideInliningLambda = inliningContext.isInliningLambda
        val reorderParametersForRootIrLambdaInvoke =
            isInsideInliningLambda &&
                    inliningContext.parent?.isInliningLambda == false &&
                    inliningContext.lambdaInfo is IrExpressionLambda
        val newArgumentList = if (reorderParametersForRootIrLambdaInvoke) {
            // Inlining a lambda. Depending on the representation, the node may or may not have the captures as
            // parameters already (`invoke` of an object, for example, does not, as captures are passed to the
            // constructor; private static methods like ones generated by the IR backend do). `invokeMethod`,
            // on the other hand, is known to never have them.
            Type.getArgumentTypes(inliningContext.lambdaInfo!!.invokeMethod.descriptor) + parameters.capturedTypes
        } else {
            Type.getArgumentTypes(node.desc) + parameters.capturedTypes
        }
```
and then...

Reply:
.... change 
```
 if (inliningContext.isInliningLambda && lambdaInfo is IrExpressionLambda) {
```
below to
```
if (reorderParametersForRootIrLambdaInvoke && lambdaInfo is IrExpressionLambda)
```
and substitute to

```private val isInliningLambda = nodeRemapper.isInsideInliningLambda```  with isInsideInliningLambda

Reply:
>and substitute `private val isInliningLambda = nodeRemapper.isInsideInliningLambda` with isInsideInliningLambda

Not sure what you mean here. Also, this variable doesn't seem relevant to the PR, so I'd prefer to avoid touching it.